src/object/obj_tx.c

```diff
@@ -212,7 +214,7 @@ dc_tx_free(struct d_hlink *hlink)
 	d_backoff_seq_fini(&tx->tx_backoff_seq);
 
 	if (tx->tx_epoch_task != NULL)
-		tse_task_decref(tx->tx_epoch_task);
+		tse_task_complete(tx->tx_epoch_task, -DER_CANCELED);
```

Comment:
complete_epoch_task() has been registered to the epoch_task, if we call tse_task_decref(), we will leak related DRAM for such CB. On the other hand, we are freeing the TX handle, under such case, epoch_task is useless any longer, cannel it via tse_task_complete() is more safe.

Reply:
I am also concerned with the new update. i don't understand why the task should be compelted with DER_CANCELLED here.
this could be the same as the IO task right? completion should be done in the cli_obj.c level in this case, no?

Reply:
When the logic comes here, the TX will be destroyed, then anything attached to the TX become useless, including this epoch task. That is why we complete it with "DER_CANCELLED". Even if related epoch task has already been scheduled (and may be completed by race), it is harmless, because tse_task_complete() logic will check whether the task is already "completed" or not.

BTW, the cli_obj and obj_tx are in the same level.

Reply:
If we do not change this decref call to a complete call, either tx_epoch_task is not the last reference, or it has already been completed---we are fine, ..., no...?

Reply:
The epoch task does not hold reference on the TX, we cannot guarantee "either tx_epoch_task is not the last reference, or it has already been completed".

On the other hand, even if tse_task_complete() has already called on the epoch task, it is still harmless to call it again, right?

Reply:
> The epoch task does not hold reference on the TX, we cannot guarantee "either tx_epoch_task is not the last reference, or it has already been completed".

Let's see, hmm...: If this is the last reference, then tx_epoch_task must have completed, or its code is buggy. If this is not the last reference, then we are fine. Which argument is wrong?

> On the other hand, even if tse_task_complete() has already called on the epoch task, it is still harmless to call it again, right?

It might be harmless to call complete again. But is it necessary to have task X complete task Y, who may be executing and not really completed (which seems to be a tricky case to support)?

Reply:
Please note that if complete_epoch_task() is called, then "tx->tx_epoch_task" will be NULL. So here if "tx->tx_epoch_task != NULL", then means it is not competed?

Reply:
OK, so decref is fine here, correct?

Reply:
> It might be harmless to call complete again. But is it necessary to have task X complete task Y, who may be executing and not really completed (which seems to be a tricky case to support)?

If there is dependency between task X and task Y, then it is normal to "have task X complete task Y", that is our tse logic does.

If we only call tse_task_decref() without tse_task_complete(), then we will leak the DRAM for complete CB. Depends on whether related epoch task has been scheduled or not. It is not impossible to give more check, but calling tse_task_complete() can avoid related check and conditions, then why not?

Reply:
If we only call tse_task_decref here, the epoch task will execute all its completion callbacks to release resources, so there shall be no memory leaks, right? The epoch task has been scheduled.

Reply:
> but calling tse_task_complete() can avoid related check and conditions, then why not?

Although I don't know tse well, I can see two usages for tse_task_complete:

- Task X completes itself.
- Task Y completes task X before scheduling it. (This is the error handling path you fixes in other places of this patch.)

Calling tse_task_complete here would be a third, different usage:

- Task Y completes task X who is executing. While this _may_ work, I think we should avoid doing so for tx_epoch_task, unless it is necessary (or if my understanding is wrong).

Reply:
> The epoch task has been scheduled.

OK, double checked current callers, related epoch tasks did scheduled. Then normally, when the epoch task completed, it will trigger related CBs and release CB DRAM. If add new callers, then need more attention on related logic.

Here, it looks more like some antinomy: normally, when the last TX reference to be released here, related task, in spite of the epoch task or other dependency tasks, should have been completed, so we may add assert "D_ASSERT(tx->tx_epoch_task == NULL);". But we cannot prevent the application aborting the TX before some tasks competed. So if we use tse_task_complete(), it will speed up the abort/cancel process as application expected; otherwise, have to wait all related epoch task and dependency tasks to complete normally. Anyway, that may be not too much important for this ticket, I will consider that in other patch to make this one to go ahead.

Reply:
> > The epoch task has been scheduled.
> 
> OK, double checked current callers, related epoch tasks did scheduled. Then normally, when the epoch task completed, it will trigger related CBs and release CB DRAM. If add new callers, then need more attention on related logic.

Yes, dc_tx_get_epoch already documents its usage like this:

    /**
     * Get the TX epoch for TX operations. See the return values.
     *
     * \param[in,out]	task	current task     <------ HERE

Hence, the epoch task shall have been scheduled.

> Here, it looks more like some antinomy: normally, when the last TX reference to be released here, related task, in spite of the epoch task or other dependency tasks, should have been completed, so we may add assert "D_ASSERT(tx->tx_epoch_task == NULL);". But we cannot prevent the application aborting the TX before some tasks competed. So if we use tse_task_complete(), it will speed up the abort/cancel process as application expected; otherwise, have to wait all related epoch task and dependency tasks to complete normally. Anyway, that may be not too much important for this ticket, I will consider that in other patch to make this one to go ahead.

Personally, I would prefer tracking and waiting for all tasks belonging to a dc_tx when restarting or aborting the dc_tx, in order to avoid accumulating unreleased resources. It's a good idea to put this aside since this PR is otherwise good.